File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.js
/**
 * @param {string} s
 * @param {string[][]} knowledge
 * @return {string}
 */
var evaluate = function(s, knowledge) {
    const mp = new Map();

    for (const [key, value] of knowledge) {
        mp.set(key, value);
    }

    let ans = "";

    for (let i = 0; i < s.length; i++) {

        if (s[i] === '(') {
            let j = i + 1;

            while (s[j] !== ')') {
                j++;
            }

            const key = s.slice(i + 1, j);

            ans += mp.has(key) ? mp.get(key) : "?";

            i = j;
        } else {
            ans += s[i];
        }
    }

    return ans;
};

File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        int n = s.size();

        string ans = "";
        int i = 0; 

        unordered_map<string, string> mpp;

        for(auto x: knowledge){

            mpp[x[0]] = x[1];
        }

        // for(auto x: mpp){
        //     cout<<x.first<<' '<<x.second;
        //     cout<<endl;
        // }


        while(i<n){
            if(s[i] == '('){
                string tmp = "";
                i++;
                while(s[i]!=')'){
                    tmp+=s[i];
                    i++;
                }

                // cout<<tmp<<endl;

                if(mpp.contains(tmp)){
                    ans+=mpp[tmp];
                } else ans+='?';
            } else ans+=s[i];

            i++;
        }

        return ans;
    }
};